tools/autograd/templates/ViewFuncs.h

```diff
@@ -0,0 +1,149 @@
+  /// Sets the values of any Tensors in the saved state. The input vector size must
+  /// match the number of Tensors in the saved state (i.e. the size of the list
+  /// returned by get_tensors()).
+  virtual void set_tensors(const std::vector<at::Tensor>&) = 0;
+};
+
+// RAII guard for modifying and restoring saved SymInt / tensor state in a
+// ViewFunc.
+struct ViewFuncSavedStateGuard {
+  ViewFuncSavedStateGuard(
+      ViewFunc& view_func,
+      const std::optional<std::vector<c10::SymInt>>& new_symints,
+      const std::optional<std::vector<at::Tensor>>& new_tensors)
+      : view_func(view_func) {
+    // save old state and set new state
+    old_symints = view_func.get_symints();
+    old_tensors = view_func.get_tensors();
+    if (new_symints.has_value()) {
+      view_func.set_symints(*new_symints);
+    }
+    if (new_tensors.has_value()) {
+      view_func.set_tensors(*new_tensors);
+    }
+  }
+
+  ~ViewFuncSavedStateGuard() {
+    // restore previous state
+    view_func.set_symints(old_symints);
+    view_func.set_tensors(old_tensors);
+  }
+
+  ViewFunc& view_func;
+  std::vector<c10::SymInt> old_symints;
+  std::vector<at::Tensor> old_tensors;
+};
+
+struct ChainedViewFunc : public ViewFunc {
+  ChainedViewFunc(
+      const std::shared_ptr<ViewFunc>&,
+      const std::shared_ptr<ViewFunc>&);
+  virtual ~ChainedViewFunc() override {};
+  virtual std::vector<c10::SymInt> get_symints() const override;
+  virtual std::vector<at::Tensor> get_tensors() const override;
+  virtual at::Tensor operator()(
+      const at::Tensor&,
+      const std::optional<std::vector<c10::SymInt>>& = c10::nullopt,
+      const std::optional<std::vector<at::Tensor>>& = c10::nullopt) const override;
+
+protected:
+  // Purposefully do nothing for these and let the sub-ViewFuncs handle it
+  virtual void set_symints(const std::vector<c10::SymInt>&) override {}
+  virtual void set_tensors(const std::vector<at::Tensor>&) override {}
+
+private:
+  std::shared_ptr<ViewFunc> first;
+  // NB: These must be mutable so we can store them during the const getter call
+  mutable size_t num_first_symints;
+  mutable size_t num_first_tensors;
+
+  std::shared_ptr<ViewFunc> second;
+  // NB: These must be mutable so we can store them during the const getter call
+  mutable size_t num_second_symints;
+  mutable size_t num_second_tensors;
+};
+
+struct ErroringViewFunc : public ViewFunc {
+  ErroringViewFunc(const std::string& error_msg) : error_msg(error_msg) {}
+  virtual ~ErroringViewFunc() override {};
+  virtual std::vector<c10::SymInt> get_symints() const override {
+    return {};
+  }
+  virtual std::vector<at::Tensor> get_tensors() const override {
+    return {};
+  }
+  virtual at::Tensor operator()(
+      const at::Tensor&,
+      const std::optional<std::vector<c10::SymInt>>& = c10::nullopt,
+      const std::optional<std::vector<at::Tensor>>& = c10::nullopt) const override {
+    TORCH_CHECK(false, error_msg);
+    return at::Tensor();
```

Comment:
fwiw, you don't have to return anything here, all our compilers are smart enough